src/csharp/Grpc.Core/ClientBase.cs

```diff
@@ -158,7 +182,7 @@ internal ClientBaseConfiguration(CallInvoker undecoratedCallInvoker, string host
 
             internal CallInvoker CreateDecoratedCallInvoker()
             {
-                return new InterceptingCallInvoker(undecoratedCallInvoker, hostInterceptor: (h) => host);
+                return undecoratedCallInvoker.Intercept(new ClientHeaderInterceptor((method, host, options) => Tuple.Create(this.host, options)));
```

Comment:
This can be refactored but I leave more changes here to a separate PR, as it might propagate and result in unrelated changes, so I will leave this change minimal.